newrelic-agent/src/main/java/com/newrelic/agent/service/logging/LogSenderServiceImpl.java

```diff
@@ -0,0 +1,585 @@
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+
+import static com.newrelic.agent.model.LogEvent.LOG_EVENT_TYPE;
+
+public class LogSenderServiceImpl extends AbstractService implements LogSenderService {
+    // Whether the service as a whole is enabled. Disabling shuts down all log events.
+    private volatile boolean forwardingEnabled;
+    // Key is the app name, value is if it is enabled - should be a limited number of names
+    private final ConcurrentMap<String, Boolean> isEnabledForApp = new ConcurrentHashMap<>();
+    // Number of log events in the reservoir sampling buffer per-app. All apps get the same value.
+    private volatile int maxSamplesStored;
+    // Key is app name, value is collection of per-transaction log events for next harvest for that app.
+    private final ConcurrentHashMap<String, DistributedSamplingPriorityQueue<LogEvent>> reservoirForApp = new ConcurrentHashMap<>();
+
+    private static final LoadingCache<String, String> stringCache = Caffeine.newBuilder().maximumSize(1000)
+            .expireAfterAccess(70, TimeUnit.SECONDS).executor(Runnable::run).build(key -> key);
+
+    public static final String METHOD = "add log event attribute";
+    public static final String LOG_SENDER_SERVICE = "Log Sender Service";
+
+    /**
+     * Lifecycle listener for log events associated with a transaction
+     */
+    protected final ExtendedTransactionListener transactionListener = new ExtendedTransactionListener() {
+        @Override
+        public void dispatcherTransactionStarted(Transaction transaction) {
+        }
+
+        @Override
+        public void dispatcherTransactionFinished(TransactionData transactionData, TransactionStats transactionStats) {
+            // Get log events from the transaction when it is finished
+            TransactionLogs data = (TransactionLogs) transactionData.getLogEventData();
+            storeEvents(transactionData.getApplicationName(), transactionData.getPriority(), data.events);
+        }
+
+        @Override
+        public void dispatcherTransactionCancelled(Transaction transaction) {
+            // Get log events from the transaction when it is canceled
+            // Even if the transaction is canceled we still want to send up any events that were held in it
+            TransactionLogs data = (TransactionLogs) transaction.getLogEventData();
+            storeEvents(transaction.getApplicationName(), transaction.getPriority(), data.events);
+        }
+    };
+
+    /**
+     * Listener to detect changes to the agent config
+     */
+    protected final AgentConfigListener configListener = new AgentConfigListener() {
```

Comment:
My Education...So why do we use anonymous classes like this throughout agent?

Reply:
We use them for various purposes. This one listens for config changes and notifies the `LogSenderServiceImpl` so that it can apply config changes dynamically.